From 49bf6296b67433982dad3bff3f3118ff33d3ef4d Mon Sep 17 00:00:00 2001
From: Keith Packard <keithp@keithp.com>
Date: Sat, 9 Sep 2023 00:58:07 -0700
Subject: [PATCH] sim/msp430: Add primitive semihosting support

Provide _exit() and putchar() implementations by trapping byte writes
of #1 to CG2.  This allows running automated tests

Signed-off-by: Keith Packard <keithp@keithp.com>
---
 sim/msp430/msp430-sim.c | 46 +++++++++++++++++++++++++++++------------
 1 file changed, 33 insertions(+), 13 deletions(-)

diff --git a/sim/msp430/msp430-sim.c b/sim/msp430/msp430-sim.c
index a0b8ddf8643..de3fd055521 100644
--- a/sim/msp430/msp430-sim.c
+++ b/sim/msp430/msp430-sim.c
@@ -1034,6 +1034,20 @@ maybe_perform_syscall (SIM_DESC sd, int call_addr)
   return 0;
 }
 
+static void
+msp430_semihost (SIM_DESC sd)
+{
+  switch(REG(12)) {
+  case 0:
+    sim_engine_halt(sd, MSP430_CPU(sd), NULL,
+		    REG(0), sim_exited, REG(13));
+    break;
+  case 1:
+    putchar(REG(13));
+    break;
+  }
+}
+
 static void
 msp430_step_once (SIM_DESC sd)
 {
@@ -1121,25 +1135,31 @@ msp430_step_once (SIM_DESC sd)
 	  && opcode->op[0].type == MSP430_Operand_Register
 	  && opcode->op[0].reg == MSR_CG
 	  && opcode->op[1].type == MSP430_Operand_Immediate
-	  && opcode->op[1].addend == 0
-	  /* A 16-bit write of #0 is a NOP; an 8-bit write is a BRK.  */
+	  /* only 8-bit writes are magic */
 	  && opcode->size == 8)
 	{
+	  /* an 8-bit write of #0 is a BRK.  */
 	  /* This is the designated software breakpoint instruction.  */
-	  PC -= opsize;
-	  sim_engine_halt (sd, MSP430_CPU (sd), NULL,
-			   MSP430_CPU (sd)->state.regs[0],
-			   sim_stopped, SIM_SIGTRAP);
-
-	}
-      else
-	{
-	  /* Otherwise, do the move.  */
-	  for (rept = 0; rept < n_repeats; rept ++)
+	  if (opcode->op[1].addend == 0)
 	    {
-	      DEST (SRC);
+	      PC -= opsize;
+	      sim_engine_halt (sd, MSP430_CPU (sd), NULL,
+			       MSP430_CPU (sd)->state.regs[0],
+			       sim_stopped, SIM_SIGTRAP);
+	      break;
+	    }
+	  else if (opcode->op[1].addend == 1)
+	    {
+	      msp430_semihost(sd);
+	      break;
 	    }
 	}
+
+      /* Otherwise, do the move.  */
+      for (rept = 0; rept < n_repeats; rept ++)
+      {
+	DEST (SRC);
+      }
       break;
 
     case MSO_addc:
-- 
2.40.1

